# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Importing Data

In [2]:
# Creating a path to project folder.

path = r'C:\Users\TanaT\(CF) Achievement 4 - Instacart Basket Analysis'

In [3]:
# Importing exported data from previous exercise (4.7).

df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_derived_variables.pkl'))

In [4]:
# Checking shape of dataframe. 

df_ords_prods_merge.shape

(32404859, 18)

In [5]:
# Checking dataframe.

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc   busiest_day    busiest_days busiest_period_of_day  
0  Mid-range product  Regular busy  Regularly busy           Most orders  
1  Mid-range product  Regular busy  Regularly busy        Average orders  
2  Mid-range product   Busiest day    Busiest days        Average orders  
3  Mid-range product  Regular busy    Slowest days           Most orders  
4  Mid-range product    Least busy    Slowest days        Average orders

# Task

### Step 2 - Find the aggregated mean of the 'order_number' column grouped by 'department_id' for the entire dataframe.

In [7]:
# Grouping data by 'department_id'.
# Applying .agg() function to the 'order_number' column to find the mean number of orders placed.
# Finding the average number of orders per user for each department ID. 

df_ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Step 3 - Analyze the result. How do the results for the entire dataframe differ from those of the subset?

This code gave us the average number of orders per user for each department id, to know which department(s) sell better.

When comparing the means for the department ids of the first million rows (subset) vs. the entire dataframe, the means appear to be similar.

Except for a few department ids, the means do not appear much different, but a t-test could be run to see if the means are significantly different.

For both the entire dataframe and the subset, the mean of department 21 was the highest (25.53 for subset).

The department ids with the biggest differences is department 10 and 21. The mean of department 10 for the entire dataframe is higher by about 1.5 points compared to the subset mean (18.68). While the mean of department 21 in the entire dataframe has over a 2 point difference compared to the subset mean (25.53) and is lower.

I think overall, it's surprising that the small subset is almost representative of the entire dataframe. There are differences between the two sets of means, but it actually gave a good sense of where the entire data is (at least when calculating the means for this particular column).

### Step 4 - Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform and loc functions.¶

In [8]:
# Creating a new column ('max_order').
# Grouping user IDs and selecting column 'order_number' to transform to find the max order number for each user (for every row).
# Putting the results of the transform function into the new column.

df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\TanaT\AppData\Local\Temp\ipykernel_71376\1995621429.py:5: FutureWarning: The provided callable <function max at 0x000001D7817E3240> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [9]:
# Fixing code.

df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform('max')

In [10]:
# Checking dataframe.

df_ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     NaN                  1          0   both   
11                     7.0                  1          0   both   
12                     7.0                  3          1   both   
13                    12.0                  4          1   both   
14                     2.0                  8          0   both   

      price_range_loc   busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regular busy  Regularly busy           Most orders   
1   Mid-range product  Regular busy  Regularly busy        Average orders   
2   Mid-range product   Busiest day    Busiest days        Average orders   
3   Mid-range product  Regular busy    Slowest days           Most orders   
4   Mid-range product    Least b

In [11]:
# Making a loyalty flag/label based on the 'max_order' column created.

df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
# Checking count of flags.

df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

Note: This shows that most of the customers fall into the 'Regular customer' category.

In [16]:
# Checking dataframe with the new column.

df_ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      NaN                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     NaN                  1          0   both   
11                     7.0                  1          0   both   
12                     7.0                  3          1   both   
13                    12.0                  4          1   both   
14                     2.0                  8          0   both   

      price_range_loc   busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regular busy  Regularly busy           Most orders   
1   Mid-range product  Regular busy  Regularly busy        Average orders   
2   Mid-range product   Busiest day    Busiest days        Average orders   
3   Mid-range product  Regular busy    Slowest days           Most orders   
4   Mid-range product    Least b

In [17]:
# Checking only columns about the user id, loyalty flags, and order number.

df_ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6
15     1494  Regular customer             7
16     1494  Regular customer             9
17     1540  Regular customer             4
18     1540  Regular customer             5
19     1540  Regular customer             6
20     1540  Regular customer             7
21     1540  Regular customer             8
22     1540  Regular customer             9
23     1540  Regular customer            14
24     1540  Regular customer            15
25     1540  Regular customer            18
26     1540  Regular customer            19
27     1540  Regular customer            20
28     1540  Regular customer            21
29     1540  Regular customer            22
30     1540  Regular customer            24
31     1540  Regular customer            25
32     1540  Regular customer            27
33     1540  Regular customer            28
34     1598      New customer             1
35     1598      New customer             2
36     1598      New customer             3
37     1647      New customer             1
38     1647      New customer             4
39     1647      New customer             6
40     1647      New customer             8
41     1647      New customer             9
42     2833      New customer             1
43     2850  Regular customer             7
44     2857  Regular customer            24
45     3010  Regular customer             5
46     3029      New customer             1
47     3395      New customer             4
48     3904      New customer             1
49     4006    Loyal customer             2
50     4006    Loyal customer             3
51     4006    Loyal customer             5
52     4006    Loyal customer             9
53     4006    Loyal customer            12
54     4122  Regular customer            11
55     4141      New customer             2
56     4343      New customer             1
57     4553      New customer             3
58     4845  Regular customer             6
59     6866      New customer             1

### Step 5 - Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category.

In [18]:
# Trying to determine whether the prices of the products purchased by loyal customers differ from those purchased by regular or new customers. 

df_ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Note:

- There could be a problem with the max price (outlier perhaps).
- All of the groups share a minimum price of 1 dollar.
- It seems like new customers purchase more expensive items on average compared to loyal and regular customers.
- Loyal customers appear to purchase more lower-cost items on average, which could be due to familiarity (knowing when items go on sale, using discounts, etc.)
- Regular customers are in the middle with an average of about $12.50 per item.

### Step 6 - The team wants to target different types of spenders in their marketing campaign. Create a spending flag for each user based on the average price across their orders.

In [19]:
# First creating a column containing the average prices of items users buy. 

df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [20]:
# Checking only specific columnns.

df_ords_prods_merge[['user_id', 'prices', 'avg_product_price']].head(15)

user_id  prices  avg_product_price
0       138     5.8           6.935811
1       138     5.8           6.935811
2       709     5.8           7.930208
3       764     5.8           4.972414
4       764     5.8           4.972414
5       777     5.8           6.935398
6       825     5.8           5.957576
7       910     5.8           6.680000
8      1052     5.8           7.162500
9      1052     5.8           7.162500
10     1379     5.8           5.813636
11     1480     5.8           5.081633
12     1480     5.8           5.081633
13     1480     5.8           5.081633
14     1494     5.8           6.803797

In [21]:
# Using conditions to create spending flags using the loc function. 

df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [22]:
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'

In [23]:
# Checking frequency of flags.

df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

Note: The total of both flags matches the shape (output 5), which means all of the data should have a label. 

In [24]:
# Checking entire dataframe.

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  reordered  _merge    price_range_loc   busiest_day    busiest_days  \
0  ...          0    both  Mid-range product  Regular busy  Regularly busy   
1  ...          1    both  Mid-range product  Regular busy  Regularly busy   
2  ...          0    both  Mid-range product   Busiest day    Busiest days   
3  ...          0    both  Mid-range product  Regular busy    Slowest days   
4  ...          1    both  Mid-range product    Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag  avg_product_price  \
0           Most orders        32  Regular customer           6.935811   
1        Average orders        32  Regular customer           6.935811   
2        Average orders         5      New customer           7.930208   
3           Most orders         3      New customer           4.972414   
4        Average orders         3      New customer           4.972414   

  spending_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  

[5 rows x 22 columns]

### Step 7 - Create an order frequency flag that marks the regularity of a user's ordering behavior according to the median in the 'days_since_prior_order' column.

In [25]:
# Frequent vs. Non-frequent customers and their ordering behavior. 
# Finding the median in 'days_since_prior_order' column for each user and grouping by user.
# Creating a column to put those results into.

df_ords_prods_merge['median_days_between_orders'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [26]:
# Checking dataframe.

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  _merge    price_range_loc   busiest_day    busiest_days  \
0  ...    both  Mid-range product  Regular busy  Regularly busy   
1  ...    both  Mid-range product  Regular busy  Regularly busy   
2  ...    both  Mid-range product   Busiest day    Busiest days   
3  ...    both  Mid-range product  Regular busy    Slowest days   
4  ...    both  Mid-range product    Least busy    Slowest days   

  busiest_period_of_day max_order      loyalty_flag avg_product_price  \
0           Most orders        32  Regular customer          6.935811   
1        Average orders        32  Regular customer          6.935811   
2        Average orders         5      New customer          7.930208   
3           Most orders         3      New customer          4.972414   
4        Average orders         3      New customer          4.972414   

   spending_flag median_days_between_orders  
0    Low spender                        8.0  
1    Low spender                        8.0  
2    Low spender                        8.0  
3    Low spender                        9.0  
4    Low spender                        9.0  

[5 rows x 23 columns]

In [27]:
# Creating frequency flags based on the median column created and the conditions listed. 

df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] > 20, 'ordering_behavior'] = 'Non-frequent customer'

In [28]:
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_between_orders'] > 10) & (df_ords_prods_merge ['median_days_between_orders'] <= 20), 'ordering_behavior'] = 'Regular customer'

In [29]:
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] <= 10, 'ordering_behavior'] = 'Frequent customer'

In [30]:
# Checking count of flags.

df_ords_prods_merge['ordering_behavior'].value_counts(dropna = False)

ordering_behavior
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

Note: Not sure why there are 5 rows without a flag. Need to investigate further. 

In [31]:
# Renaming this column to specify it is a flag column. 

df_ords_prods_merge = df_ords_prods_merge.rename(columns={'ordering_behavior': 'ordering_behavior_flag'})

In [32]:
# Checking dataframe. 

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc   busiest_day    busiest_days busiest_period_of_day  \
0  ...  Mid-range product  Regular busy  Regularly busy           Most orders   
1  ...  Mid-range product  Regular busy  Regularly busy        Average orders   
2  ...  Mid-range product   Busiest day    Busiest days        Average orders   
3  ...  Mid-range product  Regular busy    Slowest days           Most orders   
4  ...  Mid-range product    Least busy    Slowest days        Average orders   

  max_order      loyalty_flag avg_product_price spending_flag  \
0        32  Regular customer          6.935811   Low spender   
1        32  Regular customer          6.935811   Low spender   
2         5      New customer          7.930208   Low spender   
3         3      New customer          4.972414   Low spender   
4         3      New customer          4.972414   Low spender   

   median_days_between_orders ordering_behavior_flag  
0                         8.0      Frequent customer  
1                         8.0      Frequent customer  
2                         8.0      Frequent customer  
3                         9.0      Frequent customer  
4                         9.0      Frequent customer  

[5 rows x 24 columns]

In [35]:
# Checking specific columns.

df_ords_prods_merge[['user_id', 'days_since_prior_order','median_days_between_orders', 'ordering_behavior_flag']].head(60)

user_id  days_since_prior_order  median_days_between_orders  \
0       138                     3.0                         8.0   
1       138                    20.0                         8.0   
2       709                     6.0                         8.0   
3       764                     NaN                         9.0   
4       764                     9.0                         9.0   
5       777                    26.0                        11.0   
6       825                    30.0                        20.0   
7       910                    30.0                         6.0   
8      1052                    19.0                        10.0   
9      1052                    15.0                        10.0   
10     1379                     NaN                        20.0   
11     1480                     7.0                         7.0   
12     1480                     7.0                         7.0   
13     1480                    12.0                         7.0   
14     1494                     2.0                         7.0   
15     1494                     6.0                         7.0   
16     1494                    10.0                         7.0   
17     1540                     7.0                         7.0   
18     1540                     6.0                         7.0   
19     1540                    14.0                         7.0   
20     1540                     8.0                         7.0   
21     1540                     6.0                         7.0   
22     1540                    28.0                         7.0   
23     1540                     7.0                         7.0   
24     1540                     7.0                         7.0   
25     1540                     0.0                         7.0   
26     1540                     7.0                         7.0   
27     1540                     7.0                         7.0   
28     1540                     7.0                         7.0   
29     1540                     7.0                         7.0   
30     1540                     3.0                         7.0   
31     1540                    16.0                         7.0   
32     1540                     5.0                         7.0   
33     1540                     6.0                         7.0   
34     1598                     NaN                         2.0   
35     1598                    21.0                         2.0   
36     1598                     2.0                         2.0   
37     1647                     NaN                        25.0   
38     1647                    18.0                        25.0   
39     1647                     8.0                        25.0   
40     1647                    30.0                        25.0   
41     1647                    25.0                        25.0   
42     2833                     NaN                        13.0   
43     2850                     9.0                         7.0   
44     2857                    10.0                        11.0   
45     3010                     7.0                         7.0   
46     3029                     NaN                        28.0   
47     3395                    30.0                        18.0   
48     3904                     NaN                        21.0   
49     4006                     7.0                         7.0   
50     4006                     7.0                         7.0   
51     4006                     7.0                         7.0   
52     4006                     5.0                         7.0   
53     4006                     4.0                         7.0   
54     4122                     8.0                         7.0   
55     4141                    30.0                        30.0   
56     4343                     NaN                        30.0   
57     4553                    29.0                        30.0   
58     4845                    12.0                

# Exporting Data

### Step 9 - Export your dataframe as a pickle file and store it correctly in your "Prepared Data" folder.

In [36]:
# Exporting dataframe as a pickle file. 

df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'grouped_agg_orders_products_merged.pkl'))